reasoning_gym/code/codeio.py

```diff
@@ -0,0 +1,173 @@
+import gzip
+import json
+from dataclasses import dataclass
+from pathlib import Path
+from random import Random
+from typing import Any, Optional
+
+from ..factory import ProceduralDataset, register_dataset
+
+OUTPUT_PREDICTION_PROMPT_TEMPLATE = """
+You are given a question that requires some input and output variables as follows:
+
+{0}
+
+The input and output requirements are as follows:
+
+{1}
+
+Given the following input:
+
+{2}
+
+Can you predict the output without writing any code? Please think and then provide the exact output in the form of a JSON object as your final answer. The keys and values of the object should strictly match the output requirement as specified.
+
+Tip: Here is a reference code snippet for this question. You can refer to this code to guide your reasoning but not copy spans of code directly.
+
+{3}
+"""
+
+INPUT_PREDICTION_PROMPT_TEMPLATE = """
+You are given a question that requires some input and output variables as follows:
+
+{0}
+
+The input and output requirements are as follows:
+
+{1}
+
+Given the following output:
+
+{2}
+
+Can you predict a feasible input without writing any code? Please reason and put your final answer in the form of a JSON object, even if the there is only one input variable, with keys strictly matching the input variables' names as specified.
+
+Tip: Here is a reference code snippet for this question. You can refer to this code to guide your reasoning but not copy spans of code directly.
+
+{3}
+"""
+
+
+@dataclass
+class CodeIOConfig:
+    """Configuration for CodeI/O reasoning task generation"""
+
+    seed: Optional[int] = None
+    size: int = 500
+    input_prediction_probability: float = 0.5
+
+    def validate(self) -> None:
+        """Validate configuration parameters"""
+        assert 0.0 <= self.input_prediction_probability <= 1.0, "input_prediction_probability must be in [0, 1]"
+
+
+class CodeIODataset(ProceduralDataset):
+    def __init__(self, config: CodeIOConfig):
+        super().__init__(config=config, seed=config.seed, size=config.size)
+
+        self._data_path = Path(__file__).parent / "contrib/codeio/data.jsonl.gz"
```

Comment:
All data files should be stored in the https://github.com/open-thought/reasoning-gym/tree/main/reasoning_gym/data folder and path be retrieved via `get_data_file_path()` (see [data/__init__.py#L7](https://github.com/open-thought/reasoning-gym/blob/8ccf077faf30ccd8b033269c8a5be92353a303d8/reasoning_gym/data/__init__.py#L7)).